src/js/heuristicblocking.js

```diff
@@ -104,33 +113,138 @@ HeuristicBlocker.prototype = {
     let fqdn = (new URI(details.url)).host,
       origin = window.getBaseDomain(fqdn);
 
-    // if this is a main window request
+    // if this is a main window request, update tab data and quit
     if (details.type == "main_frame") {
-      // save the origin associated with the tab
-      log("Origin: " + origin + "\tURL: " + details.url);
       tabOrigins[details.tabId] = origin;
+      tabURLs[details.tabId] = details.url;
       return {};
     }
 
-    let tabOrigin = tabOrigins[details.tabId];
+    let tab_origin = tabOrigins[details.tabId],
+      self = this;
 
     // ignore first-party requests
-    if (!tabOrigin || origin == tabOrigin) {
+    if (!tab_origin || origin == tab_origin) {
       return {};
     }
 
```

Comment:
We should be able to bail here early by checking if we already saw this domain perform tracking on this site. Something like the following?
```javascript
    let firstParties = self.storage.getBadgerStorageObject('snitch_map').getItem(origin);
    if (firstParties && firstParties.indexOf(tab_origin) > -1) {
      return {};
    }
```